src/mdns/mdns.hpp

```diff
@@ -93,6 +94,68 @@ class Publisher : public MainloopProcessor
 
     typedef std::vector<TxtEntry> TxtList;
 
+    /**
+     * This structure represents information of a discovered service instance.
+     *
+     */
+    struct DiscoveredInstanceInfo
+    {
+        /**
+         * Constructor to initialize a `DiscoveredInstanceInfo` instance.
+         *
+         */
+        DiscoveredInstanceInfo(void)
+            : mPort(0)
+            , mPriority(0)
+            , mWeight(0)
+            , mTtl(0)
+        {
+        }
+
+        std::string          mName;     ///< Instance name.
+        std::string          mHostName; ///< Full host name.
+        Ip6Address           mAddress;  ///< IPv6 address.
+        uint16_t             mPort;     ///< Port.
+        uint16_t             mPriority; ///< Service priority.
+        uint16_t             mWeight;   ///< Service weight.
+        std::vector<uint8_t> mTxtData;  ///< TXT RDATA bytes.
+        uint32_t             mTtl;      ///< Service TTL.
+    };
+
+    /**
+     * This structure represents information of a discovered host.
+     *
+     */
+    struct DiscoveredHostInfo
+    {
+        /**
+         * Constructor to initialize a `DiscoveredHostInfo` instance.
+         *
+         */
+        DiscoveredHostInfo(void)
+            : mTtl(0)
+        {
+        }
+
+        std::string mHostName; ///< Full host name.
+        Ip6Address  mAddress;  ///< IP6 address.
+        uint32_t    mTtl;      ///< Host TTL.
+    };
+
+    /**
+     * This function is called to notify a discovered service instance.
+     *
+     */
+    typedef std::function<void(const std::string &aType, const DiscoveredInstanceInfo &aInstanceInfo)>
+        DiscoveredServiceInstanceCallback;
```

Comment:
nit:
```suggestion
    using DiscoveredServiceInstanceCallback =  std::function<void(const std::string &aType, const DiscoveredInstanceInfo &aInstanceInfo)>;
```

More C++11 style :)